src/common/middleware/call-api.js

```diff
@@ -17,39 +17,67 @@ export default defaults => () => next => async action => {
   validateDefaults(defaults);
   validateSettings(settings);
 
-  const [ requestType, successType, failureType ] = settings.types;
+  const requestType = settings.types ? settings.types[0] : null;
+  const successType = settings.types ? settings.types[1] : null;
+  const failureType = settings.types ? settings.types[2] : null;
   const resource = defaults.endpoint + settings.path;
   const optionsWithCsrfToken = withCsrfToken(defaults.csrfToken, settings.options);
   const createAction = createActionWith.bind({}, action);
 
   // Immediately dispatch request action
-  next(createAction({ type: requestType }));
+  if (requestType) {
+    next(createAction({ type: requestType }));
+  }
 
-  try {
-    const response = await fetch(resource, optionsWithCsrfToken);
-    await checkStatus(response);
-    const result = await response.json();
+  return fetch(resource, optionsWithCsrfToken)
+    .then((response) => {
+      return checkStatus(response).then(() => {
+        return response.json().then((result) => {
+          if (result.status !== 'success') {
+            throw getError(response, result);
+          }
 
-    if (result.status !== 'success') {
-      throw getError(response, result);
-    }
+          if (successType) {
+            return next(createAction({
+              type: successType,
+              payload: {
+                response: result
+              }
+            }));
+          }
 
-    return next(createAction({
-      type: successType,
-      payload: {
-        response: result
-      }
-    }));
-  }
-  catch(error) {
-    return next(createAction({
-      type: failureType,
-      payload: {
-        error: error
-      },
-      error: true
-    }));
-  }
+          return Promise.resolve(result);
+        });
+      }, (error) => {
+        if (failureType) {
+          return next(createAction({
+            type: failureType,
+            payload: {
+              error: error
+            },
+            error: true
+          }));
+        }
+
+        error.action = action;
+
+        return Promise.reject(error);
+      }, (error) => {
+        if (failureType) {
+          return next(createAction({
+            type: failureType,
+            payload: {
+              error: error
+            },
+            error: true
+          }));
+        }
+
+        error.action = action;
+
+        return Promise.reject(error);
+      });
+    });
```

Comment:
I've rewritten the call-api middleware to use promises instead of async/await. Source mapping for async/await code seems to be a bit flakey, I found it hard to step through this code to debug error responses handled by Fetch and problems in reducers.

I think this is a reasonable tradeoff within the scope of the middleware, I would still be happy to use async/await elsewhere in the application.

It is definitely something we should keep an eye on.